libs/fetcher.client.js

```diff
@@ -49,456 +37,278 @@ function parseResponse(response) {
-                    if (err) {
-                        debug('Syncing ' + request.resource + ' failed: statusCode=' + err.statusCode, 'info', NAME);
-                        return callback(err);
-                    }
-                    callback(null, parseResponse(response));
-                });
-            }
+    use_post = this.operation !== OP_READ || clientConfig.post_for_read;
 
-            // individual request is also normalized into a request hash to pass to api
-            requests = {};
-            requests[DEFAULT_GUID] = lodash.pick(request, CORE_REQUEST_FIELDS);
-            if (!request.body) {
-                delete requests[DEFAULT_GUID].body;
-            }
-            data = {
-                requests: requests,
-                context: this.context
-            }; // TODO: remove. leave here for now for backward compatibility
-            uri = this._constructGroupUri(uri);
-            allow_retry_post = (request.operation === OP_READ);
-            REST.post(uri, {}, data, lodash.merge({unsafeAllowRetry: allow_retry_post, xhrTimeout: this.xhrTimeout}, clientConfig), function postDone(err, response) {
-                if (err) {
-                    debug('Syncing ' + request.resource + ' failed: statusCode=' + err.statusCode, 'info', NAME);
-                    return callback(err);
-                }
-                var result = parseResponse(response);
-                if (result) {
-                    result = result[DEFAULT_GUID] || {};
-                } else {
-                    result = {};
-                }
-                callback(null, result.data);
-            });
-        },
+    if (!use_post) {
+        var getUriFn = lodash.isFunction(clientConfig.constructGetUri) ? clientConfig.constructGetUri : defaultConstructGetUri;
+        var get_uri = getUriFn.call(this, uri, this.resource, this._params, clientConfig, this.options.context);
+        if (!get_uri) {
+            // If a custom getUriFn returns falsy value, we should run defaultConstructGetUri
+            // TODO: Add test for this fallback
+            get_uri = defaultConstructGetUri.call(this, uri, this.resource, this._params, clientConfig, this.options.context);
+        }
+        if (get_uri.length <= MAX_URI_LEN) {
+            uri = get_uri;
+        } else {
+            use_post = true;
+        }
+    }
 
-        /**
-         * batch the requests.
-         * @method batch
-         * @param {Array} requests Array of requests objects to be batched. Each request is an object with properties:
-         *                             `resource`, `operation, `params`, `body`, `clientConfig`, `callback`.
-         * @return {Array} the request batches.
-         * @protected
-         * @static
-         */
-        batch : /* istanbul ignore next */ function (requests) {
-            if (!lodash.isArray(requests) || requests.length <= 1) {
-                return requests;
+    if (!use_post) {
```

Comment:
Yeah, we need this twice, in case the first check sets `use_post = true` which can happen if the GET url length is too big